src/core_plugins/vega/public/vega_editor_controller.js

```diff
@@ -0,0 +1,66 @@
+import compactStringify from 'json-stringify-pretty-compact';
+import hjson from 'hjson';
+import { Notifier } from 'ui/notify';
+import { uiModules } from 'ui/modules';
+
+const module = uiModules.get('kibana/vega', ['kibana']);
+module.controller('VegaEditorController', ($scope /*, kbnUiAceKeyboardModeService*/) => {
+
+  const notify = new Notifier({ location: 'Vega' });
+
+  return new (class VegaEditorController {
+    constructor() {
+      $scope.aceLoaded = (editor) => {
+        editor.$blockScrolling = Infinity;
+
+        const session = editor.getSession();
+        session.setTabSize(2);
+        session.setUseSoftTabs(true);
+
+        this.aceEditor = editor;
+      };
+
+      $scope.formatJson = (event) => {
+        this._format(event, compactStringify, {
+          maxLength: this._getCodeWidth(),
+        });
+      };
+
+      $scope.formatHJson = (event) => {
+        this._format(event, hjson.stringify, {
+          condense: this._getCodeWidth(),
+          bracesSameLine: true,
+          keepWsc: true,
+        });
+      };
+    }
+
+    _getCodeWidth() {
+      return this.aceEditor.getSession().getWrapLimit();
+    }
+
+    _format(event, stringify, opts) {
+      event.preventDefault();
+
+      let newSpec;
+      try {
+        const spec = hjson.parse(this.aceEditor.getSession().doc.getValue(), { legacyRoot: false, keepWsc: true });
+        newSpec = stringify(spec, opts);
+      } catch (err) {
+        // This is a common case - user tries to format an invalid HJSON text
+        notify.error(err);
+        return;
+      }
+
+      // ui-ace only accepts changes from the editor when they
+      // happen outside of a digest cycle
+      $scope.$$postDigest(() => {
```

Comment:
I know there has already been some discussion around this, but I would strongly advice against using any Angular internal methods and couldn't approve it this way.

Is there any specific reason, why we are not simply modifying the `ng-model` variable here instead of using `setValue`?

```js
$scope.vis.params.spec = newSpec;
```

Reply:
Yes, to maintain undo history within the editor. ui-ace clears the undo history every time the model is changed externally. We also went with `$$postDigest()` because it better described the intention, vs using `setTimeout(() => {}, 0)`, but a comment could also be used if you feel strongly about it.

Either way, this section of code really needs comments to describe why it's doing what it is doing.

Reply:
setValue appends value to the editor undo stack, instead of wiping it clean.  I'm ok either way with the $$postDigest -- this was suggested by @spalger earlier.

Reply:
Sorry, didn't see Spencer's comment.